src/java.base/share/classes/java/lang/foreign/MemorySegment.java

```diff
@@ -360,34 +361,48 @@
+ * slicing operation. This can be used to set the size of the zero-length segment to some desired size, so that
+ * the segment can then be accessed directly, as follows:
  *
  * {@snippet lang = java:
- * SegmentScope scope = ... // obtains a scope
- * MemorySegment foreign = someSegment.get(ValueLayout.ADDRESS, 0); // wrap address into segment (size = 0)
- * MemorySegment segment = MemorySegment.ofAddress(foreign.address(), 4, scope); // create new segment (size = 4)
- * int x = segment.get(ValueLayout.JAVA_INT, 0); //ok
+ * MemorySegment foreign = someSegment.get(ValueLayout.ADDRESS, 0); // wrap address into segment (size = Long.MAX_VALUE)
+ * foreign = foreign.asSliceUnbounded(0, JAVA_INT); // size = 4
+ * int x = foreign.get(ValueLayout.JAVA_INT, 0); //ok
+ *}
+ *
+ * Alternatively, if the size of the foreign segment is known statically, clients can associate a
+ * {@linkplain OfAddress#withTargetLayout(MemoryLayout) target layout} to the address layout used to obtain the
+ * segment. When an access operation, or a function descriptor that is passed to a downcall method handle,
+ * uses an address value layout with target layout {@code T}, the runtime will wrap any corresponding raw addresses
+ * with native segments with size set to {@code T.byteSize()}:
+ *
+ * {@snippet lang = java:
+ * MemorySegment foreign = someSegment.get(ValueLayout.ADDRESS.withTargetLayout(JAVA_INT), 0); // wrap address into segment (size = 4)
+ * int x = foreign.get(ValueLayout.JAVA_INT, 0); //ok
  *}
  *
- * Alternatively, clients can obtain an {@linkplain java.lang.foreign.ValueLayout.OfAddress#asUnbounded() unbounded}
- * address value layout. When an access operation, or a function descriptor that is passed to a downcall method handle,
- * uses an unbounded address value layouts, the runtime will wrap any corresponding raw addresses with native segments
- * with <em>maximal</em> size (i.e. {@linkplain java.lang.Long#MAX_VALUE}). As such, these segments can be accessed directly, as follows:
+ * Finally, clients can {@linkplain java.lang.foreign.MemorySegment#ofAddress(long, long, SegmentScope) obtain}
+ * a <em>new</em> native segment, with new spatial and temporal bounds, as follows:
  *
  * {@snippet lang = java:
- * MemorySegment foreign = someSegment.get(ValueLayout.ADDRESS.asUnbounded(), 0); // wrap address into segment (size = Long.MAX_VALUE)
+ * SegmentScope scope = ... // obtains a scope
+ * MemorySegment foreign = someSegment.get(ValueLayout.ADDRESS, 0); // wrap address into segment (size = 0)
+ * foreign = MemorySegment.ofAddress(foreign.address(), 4, scope); // create new segment (size = 4)
  * int x = foreign.get(ValueLayout.JAVA_INT, 0); //ok
  *}
  *
- * Both {@link #ofAddress(long, long, SegmentScope)} and {@link ValueLayout.OfAddress#asUnbounded()} are
+ * Both {@link #asSliceUnbounded(long, long)}, {@link ValueLayout.OfAddress#withTargetLayout(MemoryLayout)}
+ * and {@link #ofAddress(long, long, SegmentScope)} are
```

Comment:
There are more than two variants now:
```suggestion
 * All of {@link #asSliceUnbounded(long, long)}, {@link ValueLayout.OfAddress#withTargetLayout(MemoryLayout)}
 * and {@link #ofAddress(long, long, SegmentScope)} are
```